invertedai_cpp/invertedai/logger.cc

```diff
@@ -0,0 +1,131 @@
+    // };
+
+    void LogWriter::append_request(const std::string &req, const std::string &mode){
+        UL time_ms = this->get_current_time_in_milliseconds_();
+
+        if (mode == "location_info"){
+            this->loc_request_times_.push_back(time_ms);
+            this->loc_requests_.push_back(req);
+        }
+        else if (mode == "initialize"){
+            this->init_request_times_.push_back(time_ms);
+            this->init_requests_.push_back(req);
+        }
+        else if (mode == "drive"){
+            this->drive_request_times_.push_back(time_ms);
+            this->drive_requests_.push_back(req);
+        }
+    };
+
+    void LogWriter::append_response(const std::string &res, const std::string &mode){
+        UL time_ms = this->get_current_time_in_milliseconds_();
+
+        if (mode == "location_info"){
+            this->loc_response_times_.push_back(time_ms);
+            this->loc_responses_.push_back(res);
+        }
+        else if (mode == "initialize"){
+            this->init_response_times_.push_back(time_ms);
+            this->init_responses_.push_back(res);
+        }
+        else if (mode == "drive"){
+            this->drive_response_times_.push_back(time_ms);
+            this->drive_responses_.push_back(res);
+        }
+    };
+
+    void LogWriter::write_log_to_file(const std::string &dir_path){
+        json log;
+
+        // log["location_requests"] = this->write_iai_data_vector_to_log_<LocationInfoRequest>(this->loc_requests_);
+        // log["location_responses"] = this->write_iai_data_vector_to_log_<LocationInfoResponse>(this->loc_responses_);
+
+        log["location_requests"] = this->loc_requests_;
+        log["location_responses"] = this->loc_responses_;
+
+        log["location_request_times"] = this->loc_request_times_;
+        log["location_response_times"] = this->loc_response_times_;
+
+        // log["initialize_requests"] = this->write_iai_data_vector_to_log_<InitializeRequest>(this->init_requests_);
+        // log["initialize_responses"] = this->write_iai_data_vector_to_log_<InitializeResponse>(this->init_responses_);
+
+        log["initialize_requests"] = this->init_requests_;
+        log["initialize_responses"] = this->init_responses_;
+
+        log["initialize_request_times"] = this->init_request_times_;
+        log["initialize_response_times"] = this->init_response_times_;
+
+        // log["drive_requests"] = this->write_iai_data_vector_to_log_<DriveRequest>(this->drive_requests_);
+        // log["drive_responses"] = this->write_iai_data_vector_to_log_<DriveResponse>(this->drive_responses_);
+
+        log["drive_requests"] = this->drive_requests_;
+        log["drive_responses"] = this->drive_responses_;
+
+        log["drive_request_times"] = this->drive_request_times_;
+        log["drive_response_times"] = this->drive_response_times_;
+
+        std::string file_path = "iai_log" + std::to_string(this->get_current_time_in_milliseconds_()) + ".json";
```

Comment:
Maybe we can use `YYYY_MM_DD_HH_MM_SS_xxxx` instead? It will be easier to find the relevant log that way.

Reply:
To confirm, this will be just for the log name and keep the UNIX time for the request/response times?

Reply:
@Ruishenl what would be most useful in terms of time formatting for the request and response times? UNIX milliseconds or a human readable date down to the millsecond as Adam is suggesting for the file name?

Reply:
If i understand your code correctly, the current setup will only write to the file when the session is closed ?
Either way i think `YYYY-MM-DDTHH:MM:SS.sssZ` would be the most common one and i would use it for the request/response data, in terms of file name i think using epoch millisecond is okay since it's easier to sort.

Reply:
I think I will go with Adam's suggestion for the file title since our users will need to sift through these logs just as often as we will but just I confirm, you also want the human readable date/time for the timestamps of the request/response data?

Reply:
Yeah having human readable UTC  datatime in the data  would be better.

Reply:
Please check the date_time format I went with. @adscib I started with `YYYY_MM_DD_HH_MM_SS_xxx` but it looked bad so I changed it to `YYYY-MM-DD_HH:MM:SS:xxx`. This can be changed relatively easy so please let me know what you think.